includes/classes/sync/pull.php

```diff
@@ -540,6 +563,196 @@ protected function set_media_field_value( $destination, $post_data ) {
+			// When it is a component, the key sandwiches '_component_' so let's get the key itself
+			$content_key = explode('_component_', $key)[0];
+
+			// Check if the item has type wp-type-acf. We are assuming some items managed to skip the case check in the set_post_values function so we double check here.
+			if (isset($value['type']) && $value['type'] === 'wp-type-acf') {
+	
+				// Fetch the corresponding value from $this->item->content
+				$field_value = isset($this->item->content->{$content_key}) ? $this->item->content->{$content_key} : '';
+
+				// Check if item has subfields. If it has then it is a component from GC
+				if (isset($value['sub_fields'])) {
+	
+					// Prepare the subfield data
+					$subfield_keys = array();
+					foreach ($value['sub_fields'] as $sub_field_key) {
+						array_push($subfield_keys, $sub_field_key);
+					}
+					
+					// Let ACF add fields before rows. This order does some wonders and we need to revisit it.
+					update_field($value['field'], $component_row_data, $post_id);
+
+					$row_index = 0;
+					
+					foreach ($field_value as $subfield){
+						// Check if the number of elements in $keys and $values match
+						if (count($subfield_keys) === count(get_object_vars($subfield))) {
+							// Combine the arrays if the counts match
+							$component_row_data = array_combine($subfield_keys, get_object_vars($subfield));
+						} else {
+							// error_log("Number of keys and values don't match for array_combine()");
+							// Skip the current iteration if component_row_data is not available
+							continue;
+						}
+						add_row($value['field'], $component_row_data, $post_id);
+
+						
+						$subfield_key_id = -1;
+
+						$row_index ++;
+
+						foreach ($subfield as $subsubfield){
+							
+							$subfield_key_id ++;
+							
+							if (is_array($subsubfield)){
+								$item_key = $subfield_keys[$subfield_key_id];
+								$item = get_field_object($item_key);
+								if ($item['parent']){
+									$parent_key = $item['parent'];
+								}
+								if($item['sub_fields']){
+									$children = array();
+									foreach ($item['sub_fields'] as $child){
+										array_push($children, $child['key']);
+									}
+								}
+						
+								foreach ($subsubfield as $subsubfield_field){
+									
+									// if(empty($subsubfield_field)){
+									// 	continue;
+									// }
+									
+									if ($item['type'] == 'image'){
+										$upload_dir = wp_upload_dir();
+										$image_url = $subsubfield_field->url;
+										$image_data = file_get_contents( $image_url );
+										$filename = $subsubfield_field->filename;
+
+										if ( wp_mkdir_p( $upload_dir['path'] ) ) {
+											$file = $upload_dir['path'] . '/' . $filename;
+										}else {
+											$file = $upload_dir['basedir'] . '/' . $filename;
+										}
+
+										file_put_contents( $file, $image_data );
+
+										$wp_filetype = wp_check_filetype( $filename, null );
```

Comment:
Could we optimise this so instead of downloading the entire attachment into memory and then pushing it straight into a file, could we use `stream_copy_to_stream` to keep the memory footprint low?

For example:

```php
$fileSourceStream = fopen( $image_url, 'r' );
$filename = $subsubfield_field->filename;

if ( wp_mkdir_p( $upload_dir['path'] ) ) {
	$file = $upload_dir['path'] . '/' . $filename;
}else {
	$file = $upload_dir['basedir'] . '/' . $filename;
}

$fileDestinationStream = fopen( $file, 'w+' );

stream_copy_to_stream($fileSourceStream, $fileDestinationStream);

$wp_filetype = wp_check_filetype( $filename, null );
```